lib/ansible/plugins/connection/guestfs.py

```diff
@@ -0,0 +1,209 @@
+# Based on local.py (c) 2012, Michael DeHaan <michael.dehaan@gmail.com>
+# (c) 2013, Maykel Moya <mmoya@speedyrails.com>
+# (c) 2015, Toshio Kuratomi <tkuratomi@ansible.com>
+# (c) 2015, Michael Scherer <mscherer@redhat.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import os.path
+import guestfs
+import json
+import re
+
+from ansible.errors import AnsibleError
+from ansible.plugins.connection import ConnectionBase
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+# keep that code python2 and python3 compliant
+# we need to set $HOME so we are not left with
+# .ansible directory on /
+# also, I kept it {} free for clarity, hence the use of
+# disct instead of the short form with brackets.
+EXECUTION_SCRIPT = """
+import sys
+import json
+import os
+from subprocess import PIPE, Popen
+
+os.environ['HOME'] = '{}'
+
+cmd_file = sys.argv[1]
+if not os.path.exists(cmd_file):
+    sys.exit(1)
+
+cmd = json.loads(open(cmd_file).read())
+
+r = Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True)
+(stdout, stderr) = r.communicate()
+
+output = dict()
+output['rc'] = r.returncode
+output['stdout'] = stdout.decode('UTF-8')
+output['stderr'] = stderr.decode('UTF-8')
+
+json.dump(output, open(cmd_file + '.out', 'w'))
+"""
+
+
+class Connection(ConnectionBase):
+    ''' Libguestfs based connections '''
+
+    transport = 'guestfs'
+    # not sure how it would react, so better disable it
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin,
+                                         *args, **kwargs)
+
+        self.disk = self._play_context.remote_addr
+
+        if not os.path.isfile(self.disk):
+            raise AnsibleError("%s is not a file, aborting" % self.disk)
+
+        self.guestfs = guestfs.GuestFS(python_return_dict=True)
+
+        self.guestfs.add_drive_opts(self.disk)
+
+    def _find_python(self):
```

Comment:
The python detection is used by the connexion plugin for internal matter, ie be able to run the script whose code is in the EXECUTION_SCRIPT constant, which is a wrapper that do encode stdin/stdout as json, because guestfish do not return the 2 streams separate. 

This is separate and not related to the fact that ansible also use python. If perl was installed by default everywhere, I would use it, or if I had a good reason to believe that I could write a portable shell script doing the same without losing sanity points, I would do it.

I guess I could replace that part by a script that do execute the binary, dump stdout and stderr in 2 separate file and then get them, but this look more ugly than the current way (albeit more portable). And if I remember correctly, the python detection is needed so I can run raw or the various shell fragment in the first place. I can take a look again, because that code is not so fresh, but I am confident that if I could have avoided that, I would.